### DISCUSSION TITLE
Using non-async libraries for def endpoints

### DISCUSSION THREAD

**Discussion by adarshpunj** [zauberzeug/nicegui#1888] [Q&A] [answered]
### Question

Hi, I was going through [FAQs](https://github.com/zauberzeug/nicegui/wiki/FAQs#why-is-my-long-running-function-blocking-ui-updates) which states:

> NiceGUI (and the underlying FastAPI) are async frameworks. That means, no io-bound or cpu-bound tasks should be directly executed on the main thread but rather be "awaited". Otherwise they block execution on the main loop. 

Does it mean that we shouldn't query DB with non-sync frameworks like SQLAlchemy (unless we are using `async def` endpoints)?
Since any blocking code will still be [run](https://fastapi.tiangolo.com/async/#path-operation-functions) in an external threadpool, I believe we should be able to do this.
 
If the UI is getting blocked due to io-bound tasks, I believe we can use something like [BackgroundTasks](https://fastapi.tiangolo.com/tutorial/background-tasks/) or threading, if we're writing synchronous code.

Would love to hear more on this. Thanks.

**Comment by rodja** [comment, marked as answer]
> Does it mean that we shouldn't query DB with non-sync frameworks like SQLAlchemy (unless we are using async def endpoints)? 
Since any blocking code will still be [run](https://fastapi.tiangolo.com/async/#path-operation-functions) in an external threadpool, I believe we should be able to do this.

You can use io-blocking code like non-async SQLAlchemy, file reading or even the requests-library as long as your function decorated with `@ui.page` is not async. You already linked to the specific part in the FastAPI docs but here for reference:

> When you declare a path operation function with normal def instead of async def, it is run in an external threadpool that is then awaited, instead of being called directly (as it would block the server).

I've added this distinction to the FAQ.
